hoomd/md/NeighborList.h

```diff
@@ -536,6 +531,11 @@ class PYBIND11_EXPORT NeighborList : public Compute
     /// True if the number of bonds/angles/dihedrals/impropers/pairs has changed.
     bool m_topology_changed = false;
 
+#ifdef ENABLE_MPI
+    /// The system's communicator.
+    std::shared_ptr<Communicator> m_comm;
```

Comment:
The `Communicator` has a shared reference to the `SystemDefinition`. To make it accessible to all classes, SystemDefinition has a `weak_ptr` to the communicator (breaking the circular reference). The actual owner of the `Communicator` is the Python level `Simulation` class.

During tear down of an unused `Simulation`, Python's garbage collector might choose to destroy its reference to `Communicator` before the `NeighborList` (or any other class that connected to the `Communicator` signals). If these classes accessed `Communicator` only through the `weak_ptr`, they would find that it no longer points to an active `Communicator` when they are destructed and thus cannot disconnect their signals (e.g. `m_comm->getMigrateSignal()` cannot be called).

To prevent this, I set the classes that call `Communicator` methods to hold on to the shared pointer in order to keep the `Communicator` in memory until it is no longer needed.

Reply:
Do you need to disconnect signals to a destroyed communicator? Or are you saying the communicator would still be there but just not connected to the simulation?

Reply:
The communicator would not be there if the only reference left to it is the `weak_ptr`. It appears that nano-signal-slot disconnects all signals when the signal is deconstructed: 
https://github.com/NoAvailableAlias/nano-signal-slot/blob/bf2dd7e144fcfc2351761e27047866bb3cf48948/nano_signal_slot.hpp#L14
https://github.com/NoAvailableAlias/nano-signal-slot/blob/bf2dd7e144fcfc2351761e27047866bb3cf48948/nano_observer.hpp#L44

Reply:
That is what I would expect, so if `weak_ptr.extract == false` then we don't need to worry about disconnecting signals.

Reply:
This is try, however the subclass `NeighborListTree` needs to call `m_comm->getGhostLayerMaxWidth`. This is a required call whenever neighbor lists are used in simulations with domain decompositions and there is no reasonable fall back for cases where the `Communicator` may have been destroyed prior to this call.

I could move the `m_comm` from `NeighborList` to `NeighborListTree` but I don't see how this is any improvement over the current code.

Reply:
There should never be a case where that could occur, because that would imply that the communicator gets destroyed within the its simulation's run loop. However, this was more a query to begin with than something that should hold this back, so I will resolve the issue, and I think  this is ready to merge.